src/object/cli_ec.c

```diff
@@ -1140,13 +1140,15 @@ obj_ec_recx_reasb(daos_iod_t *iod, d_sg_list_t *sgl,
 				 * iod_size as 1 to make sgl be splittable,
 				 * server will not really transfer data back.
 				 */
-				if (iod_size == 0) {
+				if (iod_size == 0 && !punch) {
```

Comment:
since it does not need reassembly the sgl for punch anyway,   maybe just do
if (punch)
     continue
before if (!reasb_req->orr_size_fetch) {}  

So it seems clearer.